Refactor BaseEstimator's and Kernel's get_params and set_params

alegonz opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Hi there, this is a PR for #13555 .

#### What does this implement/fix? Explain your changes.
- Moved `_get_param_names` out of `BaseEstimator` as utility function now called `get_param_names_from_constructor`.
    - This is meant for reuse in `gaussian_process.kernels.Kernel` .
    - Added a `SignatureError` class to pass the signature from the utility function up to the appropriate exception message in `BaseEstimator` and `gaussian_process.kernels.Kernel`.
- Modified `gaussian_process.kernels.Kernel` to use the new utility function and remove duplication.
- Changed `GenericUnivariateSelect._make_selector` to use `get_params` instead of the new utility function so that it raises the proper error in case of a selector with inappropriate signature.
- Also refactored `get_params` to call from a new `_get_params_from` method. This is so `get_params` can be easily overridden to take a signature of, for example, the parent class in custom subclasses of client code.

#### Any other comments?
- Docstrings will be added later.
- Other that raising `SignatureError` (subclassed from `RuntimeError`) these changes don't add or change current behavior, so I haven't added any new tests. If new tests might seem necessary, please let me know.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

alegonz committed:
Moved _get_param_names out from BaseEstimator as a separate util function.

This is meant to reuse it in gaussian_process.kernels.Kernel.
Also added SignatureError to pass the signature and print the proper
error message depending on the class.

alegonz committed:
Use new get_param_names_from_constructor in gaussian_process.kernels.Kernel

alegonz committed:
Use get_params instead of _get_param_names in GenericUnivariateSelect._make_selector

jnothman commented on sklearn/base.py:
```diff
@@ -190,14 +175,17 @@ def get_params(self, deep=True):
+        try:
+            param_names = get_param_names_from_constructor(self.__class__)
+        except SignatureError as e:
+            raise SignatureError("scikit-learn estimators should always "
```
This is still more or less duplicated with kernels. Should we share this?

alegonz replied:
> This is still more or less duplicated with kernels. Should we share this?

I left that duplication to keep the error message of each class (`scikit-learn estimators...`, `scikit-learn kernels...`). I thought of 1) having the same `scikit-learn estimators...` message for both, or 2) passing a string argument to `get_param_names_from_constructor` to edit the message, but both felt awkward.

Though looking at it again, I noticed that `set_params` is duplicated too, so maybe `Kernel` could just inherit from `BaseEstimator` . Are there any historical reasons as to why it does not does so? By inheriting It could also benefit from the pickling and repr stuff.

Also, Kernels do not include any nested estimator-like parameters right? `get_params` is only shallow which seems enough, but `set_params` can handle nested parameters, why is this?

jnothman reviewed (commented):
Thanks. There's a lot going on, so I hope you can accept a slow review!

alegonz commented:
> Thanks. There's a lot going on, so I hope you can accept a slow review!

No worries! :ok_hand:

jnothman commented:
Kernel does not have fit. We can have something else that is a common base
class.

On Thu., 27 Jun. 2019, 12:03 am alegonz, <notifications@github.com> wrote:

> *@alegonz* commented on this pull request.
> ------------------------------
>
> In sklearn/base.py
> <https://github.com/scikit-learn/scikit-learn/pull/14175#discussion_r297683358>
> :
>
> > @@ -190,14 +175,17 @@ def get_params(self, deep=True):
>          params : mapping of string to any
>              Parameter names mapped to their values.
>          """
> -        out = dict()
> -        for key in self._get_param_names():
> -            value = getattr(self, key, None)
> -            if deep and hasattr(value, 'get_params'):
> -                deep_items = value.get_params().items()
> -                out.update((key + '__' + k, val) for k, val in deep_items)
> -            out[key] = value
> -        return out
> +        try:
> +            param_names = get_param_names_from_constructor(self.__class__)
> +        except SignatureError as e:
> +            raise SignatureError("scikit-learn estimators should always "
>
> This is still more or less duplicated with kernels. Should we share this?
>
> I left that duplication to keep the error message of each class (scikit-learn
> estimators..., scikit-learn kernels...). I thought of 1) having the same scikit-learn
> estimators... message for both, or 2) passing a string argument to
> get_param_names_from_constructor to edit the message, but both felt
> awkward.
>
> Though looking at it again, I noticed that set_params is duplicated too,
> so maybe Kernel could just inherit from BaseEstimator . Are there any
> historical reasons as to why it does not does so? By inheriting It could
> also benefit from the pickling and repr stuff.
>
> Also, Kernels do not include any nested estimator-like parameters right?
> get_params is only shallow which seems enough, but set_params can handle
> nested parameters, why is this?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/14175?email_source=notifications&email_token=AAATH25ZAZJQZ2WZ7OMNSG3P4NZJXA5CNFSM4H27BY62YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB4WW4BQ#discussion_r297683358>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAATH234RZMOX3HRBUL5BJLP4NZJXANCNFSM4H27BY6Q>
> .
>

alegonz commented:
>Kernel does not have fit.
Oh, right. Forgot about that. I'm now realizing that even though Estimators are conceptually defined as having a fit method,  `BaseEstimator` does not enforce this with, say, an `@abstractmethod`.

>We can have something else that is a common base class.
Right, I'll see what I can do.

alegonz committed:
Refactored get_params and set_params into a GetSetParamsMixin and
changed BaseEstimator and gaussian_process.kernels.Kernel to
inherit from it.

Also added InvalidParameterError to pass the parameter name and
print the proper error message in set_params depending on the class.

alegonz commented:
Just updated the PR with further refactoring. This time I extracted `get_params` and `set_params` into a `GetSetParamsMixin`, and changed `BaseEstimator` and `Kernel` to inherit from it.

While it looks like the pickling and repr stuff in `BaseEstimator` could also be used in `Kernel`, I decided to keep it minimal for the moment and only extract `get/set_params`.

Note that `BaseEstimator` and `Kernel` still override `get_params` and `set_params`. This it to preserve the current docstrings and error messages (also, `Kernel.get_params` is overridden to be shallow). I tried to avoid this and have a single docstring and error message for both cases, but couldn't think of a word that encompasses both estimators and kernels without being too broad. (The methods in `GetSetParamsMixin` are worded generically.)

jnothman commented on sklearn/base.py:
```diff
@@ -236,6 +203,61 @@ def set_params(self, **params):
+        """
+        try:
+            return super().get_params(deep=deep)
+        except SignatureError as e:
```
I think I would be happier with `super().__get_params(deep=deep, name='estimator')` than the introduction of these new exception classes as sentinels.

alegonz replied:
Ok, I'll remove these exception classes and pass the name as an argument.

jnothman commented on sklearn/exceptions.py:
```diff
@@ -37,6 +39,21 @@ class NotFittedError(ValueError, AttributeError):
+        self.param_name = param_name
+
+
+class SignatureError(RuntimeError):
```
I don't see why this should be public

alegonz replied:
You're right. I didn't think about that. Anyway, this will be gone as per the comment above.

alegonz committed:
Removed custom exception classes. Simplfied error messages in get_params and set_params.

alegonz commented:
I removed the custom exception classes and rephrased the error messages.

I felt that, like the custom classes I had added, introducing a `name='estimator'` argument adds noise to the signature, and complicates things while adding little value. Both approaches are kinda overkill to just choose `estimator` or `kernel` in the error message. So I instead rephrased the error message of `get_params` and `set_params` to avoid these workarounds.

alegonz commented on sklearn/base.py:
```diff
@@ -237,6 +206,54 @@ def set_params(self, **params):
+                               "specify their parameters in the signature"
+                               " of their __init__ (no varargs)."
+                               " %s doesn't follow this convention."
+                               % self.__class__) from e
```
The `from e` will show the problematic signature in the error raised by `get_param_names_from_constructor`.

alegonz commented on sklearn/base.py:
```diff
@@ -237,6 +206,54 @@ def set_params(self, **params):
+                               " %s doesn't follow this convention."
+                               % self.__class__) from e
+
+    def set_params(self, **params):
```
Left to preserve the docstring.

alegonz replied:
I would just use the inherited method instead of overriding, but I left it as I wasn't sure if there are any rules regarding docstrings.

alegonz commented on sklearn/gaussian_process/kernels.py:
```diff
@@ -126,39 +125,22 @@ def get_params(self, deep=True):
-        return params
+                               % self.__class__) from e
 
     def set_params(self, **params):
```
Left to preserve the docstring.

alegonz commented:
@jnothman Just rebased onto latest master. The `try except` with the warning on `AttributeError` was also refactored :)

jnothman commented:
Sorry I'm short on review time at the moment. Give us reasonable time to
get to this and ping if you need to

>

alegonz commented:
> Sorry I'm short on review time at the moment. Give us reasonable time to get to this and ping if you need to
> […](#)

No worries :+1: 
Just updating with the latest changes so it doesn't go stale :)

mglowacki100 commented:
@alegonz 
Hi, I'm dealing with the same issue and I checked your PR. Should I define `get_params` for my child classes or is it enough to inherit from your `BaseEstimator`? I'm asking because such drop-in replacement doesn't work for me with `**kwargs` (only parameters child class are returned).
Btw. now I use my own solution. Namely in child class I need only to define:
```
    def get_params(self, deep=True):
        """
        Get parameters.
        Based on: https://stackoverflow.com/questions/59248211
        """
        params = super().get_params(deep)
        cp = copy.copy(self)
        cp.__class__ = cp.__class__.__bases__[0]
        params.update(cp.__class__.get_params(cp, deep))
        return params
```
Subseqeuntly, all grandchildren have automagically proprer signature. However, there are a few disadvantages:
- you can use abstract class with this trick
- `__bases__[0]` makes assumptation about order in multiple inheritance
- code looks tricky

I've looked at your solution, particularly `clone` function and it seems to be based on the same trick plus a lot of checks and MixIn due to kernels. Am I right or miss something?

alegonz commented:
@mglowacki100 

Hi there. Sorry for the late reply.

>Should I define get_params for my child classes or is it enough to inherit from your BaseEstimator?

It is not enough to inherit from `BaseEstimator`. You would need to also override `get_params` to gather the parameters of its class and any super-class you might want to add _explicitly_. So for example you could do (copy-pasted from [Issue #13555](https://github.com/scikit-learn/scikit-learn/issues/13555)):

```python
class MySubclass(SomeSKLearnEstimator):
    def __init__(self, myparam=42, **kwargs):
        # Just pass **kwargs.
        # No duplicated code.
        super().__init__(**kwargs)
        self.myparam = myparam
        
    def get_params(self, deep=True):
        # Gotta override get_params. But it shouldn't be hard.
        
        # params from parent
        parent_names = self._get_param_names(super())
        params = self._get_params_from(parent_names, deep=deep)
        
        # Add my params
        names = self._get_param_names(self.__class__)
        params.update(self._get_params_from(names, deep=deep))
        # or just simply
        # params['myparam'] = self.myparam
        
        return params


x = MySubclass(myparam=42, param1=2, param2='bar')
print(x.get_params())
# {'param1': 2, 'param2': 'bar', 'myparam': 42}
```

Currently `_get_param_names(cls)`, which does the job of extracting the parameters from the `__init__` signature, is a `classmethod`, making the above hard to achieve. So, the purpose of this PR is 1) to refactor that so things like the above becomes possible, and 2) remove some duplicated code in `sklearn.gaussian_process.kernels.py`.

AFAIK, without this PR one is left with three options:
1. Your approach, which I would not suggest for the reasons you stated. Also, I would add that that approach has another disadvantage of not being explicit on which super-classes its getting the parameters from.
2. Maybe you could also hack `_get_param_names` to unbind it from the class, so you can pass to it any class you like (see [this comment](https://github.com/scikit-learn/scikit-learn/issues/13555#issuecomment-505005706)).
3. Give up and duplicate the parameters on all the sub-classes :(

>I've looked at your solution, particularly clone function and it seems to be based on the same trick plus a lot of checks and MixIn due to kernels. Am I right or miss something?

`clone` has been part of `sklearn.base` from long ago and this PR is not touching any of that.

adrinjalali commented:
@alegonz this looks pretty good, would you be able to rebase on `main` and have a fresh start here?

alegonz commented:
Hi @adrinjalali , sorry for the late reply.

Sure, I'd be happy to retake this PR. It's been a while since I opened this PR and the related issue so please give me some time to regain the context, re-study the code and address the rebase conflicts. Do you have any specific timeline for this PR?

adrinjalali commented:
@alegonz not specific timeline. Just wanted to know if we should close this or give it to somebody else. But if you're happy to have another go, that'd be perfect.
